tests/CribblyBackend.Core.UnitTests/Players/Services/PlayerServiceTests.cs

```diff
@@ -0,0 +1,153 @@
+using CribblyBackend.Core.Players.Services;
+using CribblyBackend.Test.Support;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+using System;
+using CribblyBackend.Core.Players.Repositories;
+using CribblyBackend.Core.Players.Models;
+
+namespace CribblyBackend.Core.UnitTests.Players.Services
+{
+    public class PlayerServiceTests
+    {
+        private readonly IPlayerService _playerService;
+        private readonly IPlayerRepository _fakePlayerRepository;
+        public PlayerServiceTests()
+        {
+            var provider = ServiceProviderFactory.GetProvider();
+            _playerService = provider.GetRequiredService<IPlayerService>();
+            _fakePlayerRepository = provider.GetRequiredService<IPlayerRepository>();
+        }
+
+        [Fact]
+        public async Task GetOrCreateAsyncShouldCreate_IfPlayerDoesNotExist()
+        {
+            var player = await _playerService.GetOrCreateAsync(new Player { AuthProviderId = "an auth id", Email = "abc@abc.com", Name = "james bond" });
+
+            Assert.Equal(1, player.Id);
+            Assert.Equal("an auth id", player.AuthProviderId);
+            Assert.Equal("abc@abc.com", player.Email);
+            Assert.Equal("james bond", player.Name);
+            Assert.False(player.IsReturning);
+        }
+
+        [Fact]
+        public async Task GetOrCreateAsyncShouldGet_IfPlayerDoesExist()
+        {
+            var p = await _fakePlayerRepository.CreateAsync(new Player { AuthProviderId = "an auth id", Email = "abc@abc.com", Name = "james bond" });
+
+            var player = await _playerService.GetOrCreateAsync(p);
+
+            Assert.Equal(1, player.Id);
+            Assert.Equal("an auth id", player.AuthProviderId);
+            Assert.Equal("abc@abc.com", player.Email);
+            Assert.Equal("james bond", player.Name);
+            Assert.True(player.IsReturning);
+        }
+
+        [Fact]
+        public async Task GetOrCreateAsyncShouldThrowOnDuplicateEmail_IfCreating()
+        {
+            await _fakePlayerRepository.CreateAsync(new Player { AuthProviderId = "an auth id", Email = "abc@abc.com", Name = "james bond" });
+            await Assert.ThrowsAsync<Exception>(
+                () => _playerService.GetOrCreateAsync(new Player { AuthProviderId = "an auth id1", Email = "abc@abc.com", Name = "james bond1" })
+            );
+        }
+
+        [Fact]
+        public async Task GetByIdShouldGetCorrectPlayer()
+        {
+            var testP1 = await _fakePlayerRepository.CreateAsync(
+                new Player { AuthProviderId = TestData.NewString(), Email = TestData.NewString(), Name = TestData.NewString() }
+            );
+            var testP2 = await _fakePlayerRepository.CreateAsync(
+                new Player { AuthProviderId = TestData.NewString(), Email = TestData.NewString(), Name = TestData.NewString() }
+            );
+
+            var p1 = await _playerService.GetByIdAsync(1);
+            Assert.Equal(1, p1.Id);
```

Comment:
Probably a little brittle to assert the ID like this; maybe I should use `testP1`'s fields instead